Add test for react-apollo

cometkim opened the pull request:
Hello @dai-shi, this is a pretty interesting experiment and pretty sure to have a big impact on the react ecosystems.

I'm using apollo-client in production, and I suspected the same problem since it also manages the global state (inmemory-cache) and using custom subscription internally.

And the result is:

```
  react-apollo
    check with events from outside
      ✓ check 1: updated properly (3380ms)
      ✕ check 2: no tearing during update (21ms)
      ✓ check 3: ability to interrupt render
      ✕ check 4: proper update after interrupt (5189ms)
    check with useTransaction
      ✓ check 5: updated properly with transition (3528ms)
      ✕ check 6: no tearing with transition (6ms)
      ✕ check 7: proper branching with transition (5455ms)
```

They're doing massive refactoring for v3 right now, but not sure if they've considered this.

I wanna report this to the apollo core team. Please let me know if anything goes wrong here. Thank you for the great insight.

dai-shi commented:
Hi, thanks for your interest.
I found a strange behavior probably related to your test code.
If you click "Increment shared count from outside of React (two clicks to increment one)" button twice, the count should be incremented by one. Currently, if I click the button twice, the count is plus two, instead of one.
This behavior should be resolved first before discussing further.
I haven't looked into the code yet, will do later.

Also don't be too nervous with the result. It's not a bug or any kind. It's a behavior. I can't imagine that tearing becomes a problem with most of ordinary apps.

dai-shi commented on src/react-apollo/index.js:
```diff
@@ -0,0 +1,119 @@
+
+const typeDefs = gql`
+  type Query {
+    count: Int!
```
```suggestion
    dummy: Int!
    count: Int!
```

dai-shi commented on src/react-apollo/index.js:
```diff
@@ -0,0 +1,119 @@
+        increment(root, args, { cache }) {
+          const { count } = cache.readQuery({
+            query: COUNT_QUERY,
+          });
```
Please use the reducer in common.js.
Something like:
```js
const newData = reducer({ dummy, count }, { type: 'increment' });
```

cometkim commented:
Thank you for your review @dai-shi! I fixed code to use the reducer.

I updated the test and its results.

As you say, I don't think the tearing is going to be critical, at least for my products, but I want the community to be aware of it earlier and not to be trapped. so thanks again!

dai-shi commented:
Thanks for the fix.
It's a bit unexpected to me that check 4 is failing.
Please be aware there might still be an issue in the test code that even I didn't notice.

Other than that, all seems good. Let me merge this.
Thanks for your contribution!
